FormidableLabs/react-native-app-auth
# feat/detect network errors (android only) - BREAKING CHANGE

Fixes #861 for Android only

## Description
This pull request adds error handling for network connectivity issues in the `refresh` and other function of this React Native module. Previously, the `refresh` or other function did not check for network connectivity before performing the token refresh, leading to potential issues when the device was not connected to a network. Introduced `network_error` error code for all functions. 

Though this is a useful feature, it will create a breaking change as we must follow the steps given below to use it.


### Allow permission to access the network state - BREAKING CHANGE
You need to add the `ACCESS_NETWORK_STATE` permission to your `AndroidManifest.xml` file. Here's how you can do it:

1. Open your `AndroidManifest.xml` file located in the `android/app/src/main` directory of your React Native project. 
2. Add the following line inside the `<manifest>` element, just before the `<application>` element:

```xml
<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
```
Your manifest should look something like this:

```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.yourpackagename">

    <uses-permission android:name="android.permission.INTERNET"/>
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>

    <application
        android:name=".MainApplication"
        android:label="@string/app_name"
        android:icon="@mipmap/ic_launcher"
        android:roundIcon="@mipmap/ic_launcher_round"
        android:allowBackup="false"
        android:theme="@style/AppTheme">
        ...
    </application>
</manifest>
```

## Steps to verify
1. Run the app
2. Authorize
3. Disconnect the Phone or Emulator from Internet
4. Try to use `refresh` function like below snippet.
5. It should have `network_error` in `error.code`
```javascript
 const handleRefresh = useCallback(async () => {
    try {
      const config = configs[authState.provider];
      const newAuthState = await refresh(config, {
        refreshToken: authState.refreshToken,
      });

      setAuthState(current => ({
        ...current,
        ...newAuthState,
        refreshToken: newAuthState.refreshToken || current.refreshToken,
      }));
    } catch (error) {
      Alert.alert('Failed to refresh token', error.code);
    }
  }, [authState]);
```
